File: magnus/core/connector.js
"use strict";

var Subscribable = require("../lib/utils/subscribable");
var Handler = require("../lib/wDispatcher/handler");
var String = require("../lib/wType/string");
var Address = require("../lib/wType/address");
var Uint64 = require("../lib/wType/uint64");
var Object = require("../lib/wType/object");
var Vocabulary = require("../lib/wType/vocabulary");
var Socket = require("../lib/wSocket/socket");

var Connector = Subscribable.inherit({
    "className": "Connector",
    "constructor": function(dp, srv, cmds) {
        Subscribable.fn.constructor.call(this);
        
        this._dispatcher = dp;
        this._server = srv;
        this._commands = cmds;
        this._nodes = global.Object.create(null);
        this._ignoredNodes = global.Object.create(null);
        
        this._server.on("newConnection", this._onNewConnection, this);
        this._server.on("closedConnection", this._onClosedConnection, this);
        
        var cn = new Address(["connect"]);
        var ch = new Handler(this._commands.getAddress()["+"](cn), this, this._h_connect);
        var vc = new Vocabulary();
        vc.insert("address", new Uint64(Object.objectType.String));
        vc.insert("port", new Uint64(Object.objectType.Uint64));
        this._commands.addCommand("connect", ch, vc);
        this._commands.enableCommand("connect", true);
        cn.destructor();
    },
    "destructor": function() {
        this._server.off("newConnection", this._onNewConnection, this);
        this._server.off("closedConnection", this._onClosedConnection, this);
        
        this._commands.removeCommand("connect");
        
        for (var key in this._nodes) {
            this._commands.removeCommand("disconnect" + key);
        }
        
        Subscribable.fn.destructor.call(this);
    },
    "addIgnoredNode": function(name) {
        this._ignoredNodes[name] = true;
    },
    "sendTo": function(key, event) {
        var id = this._nodes[key];
        if (!id) {
            throw new Error("An attempt to access non existing node in connector");
        }
        this._server.getConnection(id).send(event);
    },
    "_onNewConnection": function(socket) {
        var name = socket.getRemoteName().toString();
        
        if (this._ignoredNodes[name] === undefined) {
            if (this._nodes[name] === undefined) {
                if (this._server.getName().toString() === name) {
                    this.trigger("serviceMessage", "An attempt to connect node to itself, closing connection", 1);
                    setTimeout(this._server.closeConnection.bind(this._server, socket.getId()));
                } else {
                    var dc = "disconnect";
                    var dn = dc + name;
                    var dh = new Handler(this._commands.getAddress()["+"](new Address([dc, name])), this, this._h_disconnect);
                    this._commands.addCommand(dn, dh, new Vocabulary());
                    this._commands.enableCommand(dn, true);
                    
                    this._nodes[name] = socket.getId();
                    
                    this.trigger("serviceMessage", "New connection, id: " + socket.getId().toString(), 0);
                    socket.on("message", this._dispatcher.pass, this._dispatcher);
                    this.trigger("nodeConnected", name);
                }
            } else {
                this.trigger("serviceMessage", "Node " + name + " tried to connect, but connection with that node is already open, closing new connection", 1);
                setTimeout(this._server.closeConnection.bind(this._server, socket.getId()));
            }
        } else {
            this.trigger("serviceMessage", "New connection, id: " + socket.getId().toString(), 0);
            socket.on("message", this._dispatcher.pass, this._dispatcher);
        }
    },
    "_onClosedConnection": function(socket) {
        this.trigger("serviceMessage", "Connection closed, id: " + socket.getId().toString());
        
        var name = socket.getRemoteName().toString();
        if (this._ignoredNodes[name] === undefined) {
            if (this._nodes[name]) {
                this._commands.removeCommand("disconnect" + name);
                delete this._nodes[name];
                this.trigger("nodeDisconnected", name);
            }
        }
    },
    "getNodeSocket": function(key) {
        var id = this._nodes[key];
        if (!id) {
            throw new Error("An attempt to access non existing node in connector");
        }
        return this._server.getConnection(id);
    },
    "_h_connect": function(ev) {
        var vc = ev.getData();
        this._server.openConnection(vc.at("address"), vc.at("port"));
    },
    "_h_disconnect": function(ev) {
        var addr = ev.getDestination();
        var id = this._nodes[addr.back().toString()];
        if (id) {
            this._server.closeConnection(id);
        }
        
    }
});

module.exports = Connector;

File: lib/wServerUtils/connector.cpp
#include "connector.h"
#include "commands.h"

U::Connector::Connector(W::Dispatcher* dp, W::Server* srv, U::Commands* cmds, QObject* parent):
    QObject(parent),
    dispatcher(dp),
    server(srv),
    commands(cmds),
    nodes(),
    ignoredNodes(),
    controllers()
{
    connect(server, SIGNAL(newConnection(const W::Socket&)), SLOT(onNewConnection(const W::Socket&)));
    connect(server, SIGNAL(closedConnection(const W::Socket&)), SLOT(onClosedConnection(const W::Socket&)));
    
    W::String cn = W::String(u"connect");
    W::Handler* ch = W::Handler::create(commands->getAddress() + W::Address({cn}), this, &U::Connector::_h_connect);
    W::Vocabulary vc;
    vc.insert(u"address", W::Uint64(W::Object::string));
    vc.insert(u"port", W::Uint64(W::Object::uint64));
    commands->addCommand(cn, ch, vc);
    commands->enableCommand(cn, true);
}

U::Connector::~Connector()
{
    commands->removeCommand(W::String(u"connect"));
    Map::const_iterator itr = nodes.begin();
    Map::const_iterator end = nodes.begin();
    
    W::String dc = W::String(u"disconnect");
    for (; itr != end; ++itr) {
        commands->removeCommand(dc + itr->first);
    }
    
    std::multimap<W::String, C::Controller*>::const_iterator cbeg = controllers.begin();
    std::multimap<W::String, C::Controller*>::const_iterator cend = controllers.end();
    
    for (; cbeg != cend; ++cbeg) {
        C::Controller* ctrl = cbeg->second;
        if (ctrl->isSubscribed()) {
            ctrl->unsubscribe();
            ctrl->unregisterController();
        }
    }
}

void U::Connector::addIgnoredNode(const W::String& name)
{
    ignoredNodes.insert(name);
}

void U::Connector::sendTo(const W::String& name, const W::Event& event)
{
    Map::const_iterator itr = nodes.find(name);
    if (itr != nodes.end()) {
        throw new NodeAccessError(name);
    } else {
        server->getConnection(itr->second)->send(event);
    }
}

void U::Connector::onNewConnection(const W::Socket& socket)
{
    W::String name = socket.getRemoteName();
    std::set<W::String>::const_iterator ign = ignoredNodes.find(name);
    if (ign == ignoredNodes.end()) {
        Map::const_iterator itr = nodes.find(name);
        if (itr == nodes.end()) {
            if (server->getName() == name) {
                emit serviceMessage("An attempt to connect node to itself, closing connection");
                server->closeConnection(socket.getId());
            } else {
                W::String dc = W::String(u"disconnect");
                W::String dn = dc + name;
                W::Handler* dh = W::Handler::create(commands->getAddress() + W::Address({dc, name}), this, &U::Connector::_h_disconnect);
                commands->addCommand(dn, dh, W::Vocabulary());
                commands->enableCommand(dn, true);
                
                nodes.insert(std::make_pair(name, socket.getId()));
                
                emit serviceMessage(QString("New connection, id: ") + socket.getId().toString().c_str());
                connect(&socket, SIGNAL(message(const W::Event&)), dispatcher, SLOT(pass(const W::Event&)));
                
                std::multimap<W::String, C::Controller*>::const_iterator beg = controllers.lower_bound(name);
                std::multimap<W::String, C::Controller*>::const_iterator end = controllers.upper_bound(name);
                
                for (; beg != end; ++beg) {
                    beg->second->registerController(dispatcher, &socket);
                    beg->second->subscribe();
                }
                
                emit nodeConnected(name);
            }
        } else {
            emit serviceMessage(QString("Node ") + QString(name.toString().c_str()) + " tried to connect, but connection with that node is already open, closing new connection");
            server->closeConnection(socket.getId());
        }
    } else {
        emit serviceMessage(QString("New connection, id: ") + socket.getId().toString().c_str());
        connect(&socket, SIGNAL(message(const W::Event&)), dispatcher, SLOT(pass(const W::Event&)));
    }
}

void U::Connector::onClosedConnection(const W::Socket& socket)
{
    emit serviceMessage(QString("Connection closed, id: ") + socket.getId().toString().c_str());
    
    W::String name = socket.getRemoteName();
    std::set<W::String>::const_iterator ign = ignoredNodes.find(name);
    if (ign == ignoredNodes.end()) {
        Map::const_iterator itr = nodes.find(name);
        if (itr != nodes.end()) {
            std::multimap<W::String, C::Controller*>::const_iterator beg = controllers.lower_bound(name);
            std::multimap<W::String, C::Controller*>::const_iterator end = controllers.upper_bound(name);
            
            for (; beg != end; ++beg) {
                beg->second->unsubscribe();
                beg->second->unregisterController();
            }
            
            emit nodeDisconnected(name);
            commands->removeCommand(W::String(u"disconnect") + name);
            nodes.erase(itr);
        }
    }
}

void U::Connector::h_connect(const W::Event& ev)
{
    const W::Vocabulary& vc = static_cast<const W::Vocabulary&>(ev.getData());
    const W::String& addr = static_cast<const W::String&>(vc.at(u"address"));
    const W::Uint64& port = static_cast<const W::Uint64&>(vc.at(u"port"));
    server->openConnection(addr, port);
}

void U::Connector::h_disconnect(const W::Event& ev)
{
    const W::Address& addr = static_cast<const W::Address&>(ev.getDestination());
    const W::String& name = addr.back();
    
    Map::const_iterator itr = nodes.find(name);
    server->closeConnection(itr->second);
}

const W::Socket* U::Connector::getNodeSocket(const W::String& name)
{
    Map::const_iterator itr = nodes.find(name);
    if (itr == nodes.end()) {
        throw new NodeAccessError(name);
    }
    return server->getConnection(itr->second);
}

void U::Connector::registerHandler(W::Handler* handler)
{
    dispatcher->registerHandler(handler);
}

void U::Connector::unregisterHandler(W::Handler* handler)
{
    dispatcher->unregisterHandler(handler);
}

const W::Socket * U::Connector::getConnection(uint64_t p_id) const
{
    return server->getConnection(p_id);
}

void U::Connector::passThroughDispatcher(const W::Event& ev) const
{
    dispatcher->pass(ev);
}

void U::Connector::registerController(C::Controller* ctrl, const W::String& node)
{
    std::set<W::String>::const_iterator iitr = ignoredNodes.find(node);
    if (iitr != ignoredNodes.end()) {
        throw 3;    //this means you're trying to receive something from one of ignored nodes, which never going to be handled in connector, most probably it's a mistake
    }
    controllers.insert(std::make_pair(node, ctrl));
    Map::const_iterator itr = nodes.find(node);
    if (itr != nodes.end()) {
        ctrl->registerController(dispatcher, server->getConnection(itr->second));
        ctrl->subscribe();              //let's say I always need them subscribed, for now at least
    }
}

void U::Connector::unregisterController(C::Controller* ctrl, const W::String& node)
{
    bool found = false;
    std::multimap<W::String, C::Controller*>::const_iterator beg = controllers.lower_bound(node);
    std::multimap<W::String, C::Controller*>::const_iterator end = controllers.upper_bound(node);
    
    for (; beg != end; ++beg) {
        if (beg->second == ctrl) {
            found = true;           //TODO make a proper way to store 'em
            break;
        }
    }
    
    if (!found) {
        throw 4;
    }
    
    if (ctrl->isSubscribed()) {
        ctrl->unsubscribe();
        ctrl->unregisterController();
    }
    controllers.erase(beg);
}

